master/internal/rm/kubernetesrm/preemptions.go

```diff
@@ -2,103 +2,80 @@ package kubernetesrm
 )
 
-// Messages that are sent to the preemption listener.
-type startPreemptionListener struct{}
+type preemptCallbackFunc func(string)
 
 type preemptionListener struct {
-	clientSet   *k8sClient.Clientset
-	podsHandler *actor.Ref
-	namespace   string
+	cb         preemptCallbackFunc
+	syslog     *logrus.Entry
+	resultChan <-chan watch.Event
 }
 
 func newPreemptionListener(
-	clientSet *k8sClient.Clientset,
-	podsHandler *actor.Ref,
+	ctx context.Context,
 	namespace string,
-) *preemptionListener {
-	return &preemptionListener{
-		clientSet:   clientSet,
-		podsHandler: podsHandler,
-		namespace:   namespace,
-	}
-}
-
-func (p *preemptionListener) Receive(ctx *actor.Context) error {
-	switch msg := ctx.Message().(type) {
-	case actor.PreStart:
-		ctx.Tell(ctx.Self(), startPreemptionListener{})
-
-	case startPreemptionListener:
-		p.startPreemptionListener(ctx)
-
-	case actor.PostStop:
-
-	default:
-		ctx.Log().Errorf("unexpected message %T", msg)
-		return actor.ErrUnexpectedMessage(ctx)
-	}
-
-	return nil
-}
-
-func (p *preemptionListener) startPreemptionListener(ctx *actor.Context) {
+	podInterface typedV1.PodInterface,
+	cb preemptCallbackFunc,
+) (*preemptionListener, error) {
 	// Check if there are pods to preempt on startup.
-	pods, err := p.clientSet.CoreV1().Pods(p.namespace).List(
-		context.TODO(), metaV1.ListOptions{LabelSelector: determinedPreemptionLabel})
+	pods, err := podInterface.List(ctx,
+		metaV1.ListOptions{LabelSelector: determinedPreemptionLabel})
 	if err != nil {
-		ctx.Log().WithError(err).Warnf(
-			"error in initializing preemption listener: checking for pods to preempt",
-		)
-		actors.NotifyAfter(ctx, 5*time.Second, startPreemptionListener{})
-		return
-	}
-
-	for _, pod := range pods.Items {
-		ctx.Tell(p.podsHandler, PreemptTaskPod{PodName: pod.Name})
+		return nil, err
 	}
 
 	rw, err := watchtools.NewRetryWatcher(pods.ResourceVersion, &cache.ListWatch{
 		WatchFunc: func(options metaV1.ListOptions) (watch.Interface, error) {
-			return p.clientSet.CoreV1().Pods(p.namespace).Watch(
-				context.TODO(), metaV1.ListOptions{LabelSelector: determinedPreemptionLabel})
+			options.LabelSelector = determinedPreemptionLabel
+			return podInterface.Watch(ctx, options)
 		},
 	})
 	if err != nil {
-		ctx.Log().WithError(err).Warnf("error initializing preemption watch")
-		actors.NotifyAfter(ctx, 5*time.Second, startPreemptionListener{})
-		return
+		return nil, err
 	}
 
-	ctx.Log().Info("preemption listener is starting")
-	for e := range rw.ResultChan() {
+	// Log when pods are first added to the informer (at start-up).
+	syslog := logrus.WithFields(logrus.Fields{
+		"component": "preemptionListener",
+		"namespace": namespace,
+	})
+	for i := range pods.Items {
+		syslog.Debugf("preemption listener added: %s", pods.Items[i].Name)
+		cb(pods.Items[i].Name)
+	}
+
+	return &preemptionListener{
+		cb:         cb,
+		syslog:     syslog,
+		resultChan: rw.ResultChan(),
+	}, nil
+}
+
+func (p *preemptionListener) run() {
```

Comment:
these listeners are all so similar now, it seems like we could shave a lot of code by refactoring them all. maybe file a ticket?

Reply:
sure thing -- I think we could honestly condense all 4 informers into one file & fewer functions.
(Note to carolina: look into making run() take in a context)